core/trino-main/src/main/java/io/trino/server/NodeStateManager.java

```diff
@@ -138,20 +197,80 @@ private void shutdown()
         catch (ExecutionException e) {
             log.warn(e, "Problem stopping the life cycle");
         }
-
         shutdownAction.onShutdown();
     }
 
+    private void drain()
+    {
+        if (nodeState.get() == DRAINING) {
+            waitActiveTasksToFinish();
+        }
+        drainingComplete();
+    }
+
+    private void drainingComplete()
+    {
+        boolean success = nodeState.compareAndSet(DRAINING, DRAINED);
+        if (success) {
+            log.info("NodeState: DRAINED, server can be safely SHUT DOWN.");
+        }
+        else {
+            log.info("NodeState: " + nodeState.get() + ", will not transition to DRAINED");
+        }
+    }
+
+    private void waitActiveTasksToFinish()
+    {
+        // At this point no new tasks should be scheduled by coordinator on this worker node.
+        // Wait for all remaining tasks to finish.
+        while (isTerminating()) {
+            List<TaskInfo> activeTasks = getActiveTasks();
+            log.info("Waiting for " + activeTasks.size() + " active tasks to finish");
+            if (activeTasks.isEmpty()) {
+                break;
+            }
+
+            waitTasksToFinish(activeTasks);
+        }
+
+        // wait for another grace period for all task states to be observed by the coordinator
+        if (isTerminating()) {
+            sleepUninterruptibly(gracePeriod.toMillis(), MILLISECONDS);
+        }
+    }
+
+    private void waitTasksToFinish(List<TaskInfo> activeTasks)
+    {
+        CountDownLatch countDownLatch = new CountDownLatch(activeTasks.size());
+
+        for (TaskInfo taskInfo : activeTasks) {
+            sqlTaskManager.addStateChangeListener(taskInfo.taskStatus().getTaskId(), newState -> {
+                if (newState.isDone()) {
+                    countDownLatch.countDown();
+                }
+            });
+        }
+
+        try {
+            countDownLatch.await();
```

Comment:
this will not be released if you "reactivate" node. 
A clean soluton would enforce that there is no longer thread hanging in `waitTasksToFinish` if we reactivate the node.

Reply:
Yes, you make a good point. The task and thread will still wait. The risk of using too many resources is low, since reactivation is  usually rare, but I can improve it (one can not expect how it will be used). 

I can replace the `countDownLatch await()` with:
```
            while (!countDownLatch.await(1, TimeUnit.SECONDS)) {
                if (!isShuttingDownOrDraining()) {
                    log.info("Wait for tasks interrupted, worker is no longer draining.");

                    break;
                }
            };
```
This will solve the issue of thread waiting. It does not cleanup the `stateChangeListeners` in sqlTaskManager, so there is still a reference to countDownLatch. But the countDownLatch is effectively final so no additional context should attached. And all of this will be cleaned when the task finishes. 

WDYT?

Reply:
This is active wait - so no pretty. But given fact this is not a common code path I a fine with that.

Reply:
Yes, no event, no interrupt just wait in a loop. But as you noticed - not a common code path. It is a shutdown/drain. 
I assume I can use it.